File: examples/interface/ZDependenceSearch.py
import El

n=40
tag=El.zTag
progress=False
timeLLL=False
outputAll=False
outputCoeff=False

# Initially draw z out of a uniform ball of radius five about 10+0i
z = El.Matrix(tag)
El.Uniform(z,n,1,10.,5.)

# Compute a (hidden) Gaussian integer vector aHidden
aHidden = El.Matrix(tag)
El.Uniform(aHidden,n-1,1,0.,5.)
El.Round(aHidden)

# Force the last entry of z to be the above integer linear combination of the
# first n-1 entries
zetaLast=El.Dotu(aHidden,z[0:n-1,:])
z.Set(n-1,0,zetaLast)

if outputAll or outputCoeff:
  El.Print( aHidden, "aHidden" )
  El.Print( z, "z" )

ctrl=El.LLLCtrl_d()
ctrl.progress = progress
ctrl.time = timeLLL

NSqrt=1000000.

# NOTE: The coefficients become orders of magnitude higher for 'weak'
#       reductions and seem to reliably lead to an exception being thrown,
#       so these tests will only use strong LLL reductions

B=El.Matrix()
for presort, smallestFirst in (True,True), (True,False), (False,False):
  for deltaLower in 0.5, 0.75, 0.95, 0.98, 0.99:
    for variant in El.LLL_NORMAL, El.LLL_DEEP, El.LLL_DEEP_REDUCE:

      print('variant={}, presort={}, smallestFirst={}, deltaLower={}'.format( \
        variant,presort,smallestFirst,deltaLower))

      ctrl.delta = deltaLower
      ctrl.variant = variant
      ctrl.presort = presort
      ctrl.smallestFirst = smallestFirst

      # Search for the linear dependence
      startTime = El.mpi.Time()
      numExact, B, U = El.ZDependenceSearch(z,NSqrt,ctrl)
      runTime = El.mpi.Time() - startTime
      print('  runtime: {} seconds'.format(runTime))
      print('  num \"exact\": {}'.format(numExact))
      if outputAll:
        El.Print( B, "B" )
        El.Print( U, "U" )
      elif outputCoeff:
        El.Print( U[:,0], "u0" )
      print('approximate zero: {}+{}i'.format( \
        B.GetRealPart(n,0)/NSqrt,B.GetImagPart(n,0)/NSqrt))

El.Finalize()

File: examples/lattice/ZDependenceSearch.cpp
/*
   Copyright (c) 2009-2016, Jack Poulson
   All rights reserved.

   This file is part of Elemental and is under the BSD 2-Clause License, 
   which can be found in the LICENSE file in the root directory, or at 
   http://opensource.org/licenses/BSD-2-Clause
*/
#include "El.hpp"
using namespace El;

#ifdef EL_HAVE_MPC
typedef BigFloat F;
#elif defined(EL_HAVE_QUAD)
typedef Quad F;
#else
typedef double F;
#endif
typedef Base<F> Real;

int main( int argc, char* argv[] )
{
    Environment env( argc, argv );

    try
    {
        const Int n = Input("--n","problem dimension",100);
        const Int varInt = Input("--variant","0: weak, 1: normal, 2: deep insertion, 3: deep reduction",2);
        const bool progress = Input("--progress","print progress?",false); 
        const bool time = Input("--time","time LLL?",false);
        const bool printAll = 
          Input("--printAll","output all matrices?",false);
        const bool printCoeff =
          Input("--printCoeff","output coefficients?",false);
        const Real NSqrt = Input("--NSqrt","sqrt of N",Real(1e6));
#ifdef EL_HAVE_MPC
        const mpfr_prec_t prec =
          Input("--prec","MPFR precision",mpfr_prec_t(256));
#endif
        ProcessInput();
        PrintInputReport();

#ifdef EL_HAVE_MPC
        mpc::SetPrecision( prec );
#endif

        // Initially draw z out of a uniform ball
        Matrix<F> z;
        Uniform( z, n, 1, F(10), Real(5) );

        // Compute a (hidden) Gaussian integer vector, aHidden
        Matrix<F> aHidden;
        Uniform( aHidden, n-1, 1, F(0), Real(5) );
        Round( aHidden );

        // Force the last entry of z to be the integer linear combination
        // of the first n-1 entries specified by aHidden
        const F zetaLast = Dotu( aHidden, z(IR(0,n-1),ALL) );
        z.Set( n-1, 0, zetaLast );

        if( printAll || printCoeff )
        {
            Print( aHidden, "aHidden" );
            Print( z, "z" );
        }

        LLLCtrl<Real> ctrl;
        ctrl.variant = static_cast<LLLVariant>(varInt);
        ctrl.progress = progress;
        ctrl.time = time;

        vector<Real>
          deltaLowers{Real(0.5),Real(0.75),Real(0.95),Real(0.98),Real(0.99)}; 
        for( Real deltaLower : deltaLowers )
        {
            Output("deltaLower=",deltaLower); 
            ctrl.delta = deltaLower;

            // Search for the linear dependence
            double startTime = mpi::Time();
            Matrix<Real> B, U;
            Int numExact = ZDependenceSearch( z, NSqrt, B, U, ctrl );
            double runtime = mpi::Time() - startTime;
            const Real oneNormBasis = OneNorm( B );
            const Real infNormUnimod = InfinityNorm( U );
            Output("  runtime: ",runtime," seconds");
            Output("  num \"exact\": ",numExact);
            Output("  approximate zero: ",B.Get(n,0)/NSqrt);
            Output("  || B ||_1 = ",oneNormBasis);
            Output("  || U ||_oo = ",infNormUnimod);
            if( printAll )
            {
                Print( B, "B" );
                Print( U, "U" );
            }
            else if( printCoeff )
            {
                Print( U(ALL,IR(0)), "u0" );
            }
        }
    }
    catch( std::exception& e ) { ReportException(e); }
    return 0;
}
